zio-http/shared/src/main/scala/zio/http/internal/QueryParamEncoding.scala

```diff
@@ -17,12 +17,311 @@
+
+    val result = new StringBuilder(end - start) // Pre-allocate with estimated size
+    i = start
+
+    while (i < end) {
+      val c = component.charAt(i)
+      if (c == '%' && i + 2 < end) {
+        // Extract hex digits directly without substring
+        val digit1 = Character.digit(component.charAt(i + 1), 16)
+        val digit2 = Character.digit(component.charAt(i + 2), 16)
+
+        if (digit1 >= 0 && digit2 >= 0) { // Valid hex digits
+          val decoded = (digit1 << 4) | digit2
+          result.append(decoded.toChar)
+          i += 3
+        } else {
+          // Invalid hex encoding, treat % as literal
+          result.append('%')
+          i += 1
+        }
+      } else if (c == '+') {
+        result.append(' ')
+        i += 1
+      } else {
+        result.append(c)
+        i += 1
+      }
+    }
+
+    // Apply charset conversion only once at the end
+    new String(result.toString.getBytes(charset))
+  }
+
+  /**
+   * Encodes a URL component according to HTML/URL encoding rules. Spaces become
+   * '+', and special characters become percent-encoded.
+   */
+  private def encodeComponent(component: String, charset: Charset): String = {
+    if (component.isEmpty) return component
+
+    // Fast path for strings that don't need encoding
+    var needsEncoding = false
+    var i             = 0
+    val len           = component.length
+    while (i < len && !needsEncoding) {
+      val c = component.charAt(i)
+      // RFC 3986 unreserved characters plus '*' (Netty-specific addition)
+      needsEncoding = !(
+        (c >= 'a' && c <= 'z') ||
+          (c >= 'A' && c <= 'Z') ||
+          (c >= '0' && c <= '9') ||
+          c == '-' || c == '.' || c == '_' || c == '~' || c == '*'
+      )
+      i += 1
+    }
+
+    if (!needsEncoding) return component
+
+    // Special optimization for UTF-8 (most common case)
+    val isUtf8 = charset == java.nio.charset.StandardCharsets.UTF_8 || charset.name().equalsIgnoreCase("UTF-8")
+
+    // Calculate a more accurate capacity based on where encoding becomes necessary
+    val estimatedCapacity =
+      if (i > 1) {
+        // First (i-1) chars don't need encoding + remaining chars might need encoding (worst case: 3x)
+        (i - 1) + ((component.length - i + 1) * 3)
+      } else {
+        // Worst case: all characters need percent encoding
+        component.length * 3
+      }
+
+    val result = new java.lang.StringBuilder(estimatedCapacity)
+
+    // Copy the characters that we know don't need encoding
+    if (i > 1) {
+      result.append(component, 0, i - 1)
+    }
+
+    // Process remaining characters
+    var j = if (i == 0) 0 else i - 1
+
+    if (isUtf8) {
+      // Optimized UTF-8 path
+      while (j < len) {
+        val c = component.charAt(j)
+        if (
+          (c >= 'a' && c <= 'z') ||
+          (c >= 'A' && c <= 'Z') ||
+          (c >= '0' && c <= '9') ||
+          c == '-' || c == '.' || c == '_' || c == '~' || c == '*'
+        ) {
```

Comment:
I think this is also used above. Maybe we can put it into a method? The JVM will be able to JIT-compile it and inline it